spec/src/main/asciidoc/XX-config-mapping.adoc

```diff
@@ -0,0 +1,192 @@
+
+- Needs to backed by a Proxy or some other means to provide an implementation
+- Harder to use standalone
+
+Some Jakarta EE specifications (JPA, JSONB) use POJOs to bind data. While similar, it is not intended for the
+end-user to make modifications directly on the configuration representation. The Config instance will handle any change
+(even with mutability support), so a POJO is not required to support the configuration needs fully.
+
+An `interface` to represent configuration complex types seems to be a better option in this case.
+
+== Records
+
+TODO
+
+== Mapping (Binding) Rules
+
+A complex object type uses the following rules to map (or bind) configuration values to their member values:
+
+- A configuration path is built by taking the object type prefix (or namespace) and the mapping (binding) member name
+- The member name is converted to its kebab-case format
+- If the member name is represented as a getter, the member name is taken from its property name equivalent, and then
+converted to its kebab-case format.
+- The configuration value is automatically converted to the member type
+- The configuration path is required to exist with a valid configuration value or the mapping (binding) will fail.
+
+=== Nested Types
+
+- A nested type contributes with its name (converted to its kebab-case format)
+- The configuration path is built by taking the root object type prefix (or namespace), the nested
+type name and the member name of the nested type
+
+=== Collections and Arrays
+
+- A member with a `Collection` or `Array` type requires the configuration name to be in its indexed format
+- Each configuration name, plus its index maps (or binds) the configuration value to the corresponding `Collection` or
+`Array` element in the object type
+- TODO - Define how to represent the indexed format. A well accepted representation is the presence of square brackets
+with the collection / array index inside after the name.
+- The index specified in the configuration name is used to order the element in the `Collection` or `Array`. Missing
+elements or gaps are removed.
+
+=== Maps
+
+- A member with a `Map` type requires an additional configuration name added to the configuration path of the `Map`
+member to act as a map key
+- The additional configuration name maps (or binds) a Map entry with the configuration name as the `Map` entry key and
+the configuration value as the Map entry value
+
+=== Optionals
+
+- A mapping (binding) can wrap any complex type with an `Optional`
+- `Optional` mappings (bindings) do not require the corresponding configuration path and value to be present
+
+== Override Conventions
+
+It is possible to override:
+
+- The base configuration path (prefix or namespace) of the mapping (binding)
+- Mapping (Binding) member names
+- The default Converter of a member
+- Add a default value to a member
+
+== Examples
+
+=== Mapping (or Binding) definition
```

Comment:
(I apologize in advance since I had to miss the last couple of meetings; perhaps this was discussed.)  Have you considered Jackson's `databind` and `dataformat-text` projects?  There are many, many edge cases which it handles that I don't see represented here.  If we're going to define a standard in this space for some reason (I disagree, for the record, that we need to), then we should look at the king in the space for inspiration, since it's some of the most widely used Java software in the world.  (Note, of course, that Jackson databind works with JSON, YAML, TOML, properties files and XML in the same fashion, using the same underlying machinery…almost like a standard 😄 )

Reply:
Hi @ljnelson. No need to apologize.

This is just the initial understanding. Of course, it needs to be expanded to add extra detail.

> If we're going to define a standard in this space for some reason (I disagree, for the record, that we need to)

What do you think we should do instead?

Reply:
Think of config like `slf4j` logging: there's a façade that a class developer uses, and then there are possible implementations under it.  The class developer interacts with the façade, not with the underlying implementation (which is something like logback, `java.util.logging`, `log4j`, whatever).  In `slf4j`'s case, there can be only one of these logging implementations at a time.

So: I think we should be focusing on "`scf4j`" (simple config façade 4 java), not on whatever its backend might be (which is what we're currently doing, and because we haven't defined the façade yet, we don't know how the backend will be used).

There are plenty of existing battle-tested object binding frameworks out there that do whatever they do.  If you can just adapt those implementations to the façade we define then you're done.  The backends already know how to do what they do.  The standard ends up doing what standards should do: adapting existing implementations to common usage patterns.

I recognize I am in the extreme minority in this committtee with this view.

Reply:
I'm fine if we do it to allow using a binding implementation instead. In reality, nothing right now stops an implementation to do it, we are just missing the facade.

Reply:
(I'm sorry, I didn't understand your comment.)  My point is, the following _kind_ of "façade" API is all that needs to be standardized:

```java
List<String> strings =
  ourConfigFacadeInterface.load(typeTokenRepresentingListString,
                                addressingInformationPartiallyIdentifyingWhatToLoad);
```

(I am not actually proposing the API above; it has many flaws; I've [sketched an actual proposal elsewhere](https://www.eclipse.org/lists/config-dev/msg00161.html). But it is easy to understand at a first glance.)

This is the `slf4j` part.  The gunk that makes the `List<String>` above is _deliberately_ standardized, just like logback is not standardized, `java.util.logging` is not standardized, `log4j` is not standardized and `commons-logging` is not standardized.  I can implement the API above using any combination of Jackson, Java properties, JNDI, whatever.  I don't need special annotations or a specification uniting them all to do so.

All of this to say: _How_ the `List<String>` is built, and by what, does not need to be standardized.  I recognize, however, that this is exactly what this PR is doing.  To be perfectly clear: I disagree with the need for this PR, and with the need for standardizing binding that it addresses.  (I also don't want to hijack this PR further so I'll stop here.)

Reply:
Please, keep the discussion going. This is an excellent discussion, and now is the right time to do it.

Again, I don't oppose the idea of leaving this out entirely of the specification and providing only a facade for the object retrieval. I think we still need to define what must be supported (collections, optionals, maps), and I see two possible downsides:

- The binding implementation may not be able to use Config to configure itself (probably not a big deal)
- A portable way to override rules 

For instance, picking up on the Jackson example, Jackson can override its conventions to do the binding (as most of all the other binding frameworks). How do you propose to achieve this in a portable way? Or you don't think it is relevant to offer such a feature? If users are aware of the backing implementation, they can still fall back to the specific APIs to override these rules in exchange for non-portable behavior. Is this acceptable?

Reply:
I don't think it is _practical_ to try to standardize object binding.  If you nevertheless want to standardize object binding, then you absolutely should start from JAXB, look at the deserialization side, and change all occurrences of "`XML`" in its binding annotations to something more generic.  Boom; you're done.  JAXB already solved this problem in a way that actually has very little to do with XML.  Quite well, in fact.

Then you'll need to add in all the extras that Jackson did in JSON-land to make sure you're covering those (deserialization) cases too, changing all occurrences of "`Json`" to something more generic.  Jackson already solved this problem in a way that actually has very little to do with JSON.

And after you do all of that, you will _still_ not be interoperable with the popular binding and configuration products out there in the world.  So why go through this work?

The lowest common denominator for object binding, supported (and in some cases required) by Spring, Jackson, Typesafe Config, JAXB, etc. etc. is still, perhaps sad to say, Java Beans conventions.  That seems good enough, since, after all, it's what all popular configuration and object bindings use as their common denominator.

Before I get thrown out of the building for saying all that 😄 , consider also there are three broad cases, all quite different, loosely dubbed "object binding" in configuration systems:

1. **Provider binds.**  I have a POJO class, and I ask a loader for an instance of it.  Let's say the loader implementation in this case happens to be backed by multiple providers (it doesn't have to be).  The system responds by asking its providers to see if any one of them can deliver such an object.  Let's say there is a Jackson provider, and it says yep, I can do that.  So the Jackson provider delivers me back a POJO instance however it does it.  If my POJO class doesn't have annotations but it's basically a valid Java bean, Jackson will work fine.  If my POJO class has Jackson annotations, then Jackson will "see" them and bind it appropriately.  If my POJO class has TurboConfig™ annotations on it (deliberately made up), then Jackson won't see them, but maybe can bind anyway, if my POJO is a Java bean.  And so on.
2. **System assembles.**  I have a POJO class, and I ask a loader for an instance of it.  The loader begins by introspecting my class and deriving the individual properties that make it up.  It then asks its providers for these properties and installs their results into a new instance of my POJO class.  Presumably the providers are deemed to be at some sort of hazy "lower level" than this machinery, i.e. no provider is going to _also_ go through this process.  Then the system does some kind of centralized conversion process on each property to adapt it to my POJO class' requirements.  Finally, my POJO instance is built reflectively and returned.
3. **POJO pulls.**  I have a POJO interface (let's say; you could do this with a class but as you'll see it could be too complicated) with various `get` methods on it.  I ask the loader for an instance of it.  The system responds by asking its providers to see if one of them can deliver such an object.  Let's say there is a provider that knows how to implement (proxy, generate, whatever) all of the `get` methods so that they are backed by relevant calls to loaders.  It returns an instance of my POJO interface, where the `get` methods all call `loader.load` appropriately, which is now effectively a "tree structure" but with the added benefit that it has documentation and schema support (javadoc and method signatures) and default value and error handling support (`default` interfaces etc.).

Case 1 does not _require_ any standardized binding mechanism other than a descriptive one ("make your object conform to the Java Beans standard"), and, if you _happen_ to annotate your POJO with annotations that a given provider "knows about" (say, Jackson), then hooray for you.  (Also, perhaps, note Jackson's concept of mixin annotations which keeps your POJO annotation-free while still permitting the use of Jackson annotations.)

Case 2 is in my opinion kind of the worst of all possible worlds and certainly does not require standardized binding annotations (Spring properties demonstrates this).  (I think we, as a committee are strangely focused on this case, and to the extent that is true, I disagree strongly with this focus.)

Case 3 is a way of building a POJO out of other dynamically fetched config objects, across providers, while permitting changes, or not, in the underlying values and also documenting things properly and indicating what is required and what is optional.  It also doesn't require any additional binding indicators (because basically true binding isn't used&mdash;no provider is going to call "setters" on it or try to set fields).

Finally, consider the end game (I know this specification is still in its infancy, but bear with me).  Suppose you _do_ come up with a suite of annotations that define some parts of object binding (but, I'm going to say, you end up not doing everything that Jackson does, because the edge-case space is absolutely enormous and it took Jackson ten years to get to its well-earned position of supremacy).  Now you want to let Jackson play in this world.  What do you do?  I'm going to guess that just as Jackson didn't (for very good reasons!) implement JSON-B, they're not going to implement this either.  So now you have to write some kind of adapter that will introspect a class and translate all of the standard annotations it finds into some kind of deserialization instructions that Jackson can support.  To gain what, exactly?  (Case 1 works without this; case 3 works without this.)  Maybe it would have been easier to just say: honor the Java Beans contract, full stop; that's the standard way to do it?  And if there are Jackson annotations on the POJO, cool, Jackson will read them?

Next, records and value types are coming.  (They're here, but they're not really here yet. 😄 )  Existing configuration and binding frameworks are going to start to adopt them and their (naming, constructor) conventions.  If we've defined object binding descriptively only ("follow conventions" rather than "use this annotation"), then we can adapt to this changing world nimbly.

Finally finally 😄 , it's not just Jackson or JAXB that says: the lowest common denominator is Java Beans/conventions. Consider [Typesafe Config's `ConfigBeanFactory`, which doesn't allow you to customize bean creation at all](https://javadoc.io/doc/com.typesafe/config/latest/com/typesafe/config/ConfigBeanFactory.html).  Or Spring properties, the most popular configuration machinery on the planet, which uses `@ConfigurationProperties` mostly as a marker annotation to identify objects to be bound&mdash;their binding normally requires them to conform to Java Beans standards.  Or `javax.sql.DataSource`s, whose specification mandates that to be configured they must conform to Java Beans conventions.  Or the old Java EE Connector specification, which says the same thing.  Or, or, or….

Finally, we go back to the façade (the loader): the contract with the _class developer_ is simply: give me a class (or ideally some rich type information) and a name (or ideally some rich addressing information), and I will try my hardest to return a suitable instance for you (or, ideally, a `Supplier`-like object for that instance).  Maybe something can bind your POJO, and maybe something can not.  Maybe your POJO ends up being some kind of synthetic object, and maybe it does not.  Maybe your POJO is something that internally uses more loader calls and maybe it does not.  You don't care.  Binding isn't really part of this contract at all.

So: for all these reasons, at a high level I disagree with the desire to define object binding (other than descriptively, relying on existing or future conventions of records or beans) as part of this specification.